In [50]:
import numpy as np
import  matplotlib.pyplot as plt
%matplotlib notebook
%matplotlib inline
import pandas as pd
import scipy as sp
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import colorama
from colorama import Fore, Back, Style
import seaborn as sns
from sklearn.model_selection import GridSearchCV
import random
from sklearn.preprocessing import StandardScaler
random.seed(0)

In [51]:
data = pd.read_csv('C:/Users/anvit/OneDrive/Documents/2nd sem docs/Scripting/MiniProject/loan_sanction.csv')

In [52]:
data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [73]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 480 entries, 1 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            480 non-null    object 
 1   Gender             480 non-null    object 
 2   Married            480 non-null    object 
 3   Dependents         480 non-null    object 
 4   Education          480 non-null    object 
 5   Self_Employed      480 non-null    object 
 6   ApplicantIncome    480 non-null    int64  
 7   CoapplicantIncome  480 non-null    float64
 8   LoanAmount         480 non-null    float64
 9   Loan_Amount_Term   480 non-null    float64
 10  Credit_History     480 non-null    float64
 11  Property_Area      480 non-null    object 
 12  Loan_Status        480 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 52.5+ KB


In [53]:
data.isnull().sum()
#checking for null values in the columns

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [54]:
#data['Gender'].fillna(0, inplace=True)
#data['Married'].fillna(0, inplace=True)
#data['Dependents'].fillna(0, inplace=True)
#data['Self_Employed'].fillna(0, inplace=True)
#data['LoanAmount'].fillna(0, inplace=True)
#data['Loan_Amount_Term'].fillna(0, inplace=True)
#data['Credit_History'].fillna(0, inplace=True)

data.dropna(inplace=True)
data.isnull().sum()
#replacing the missing values with 0

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

In [75]:
data.describe()
#describe function is used to calculate the summary statistics in Python.

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       480.000000         480.000000  480.000000        480.000000   
mean       5364.231250        1581.093583  144.735417        342.050000   
std        5668.251251        2617.692267   80.508164         65.212401   
min         150.000000           0.000000    9.000000         36.000000   
25%        2898.750000           0.000000  100.000000        360.000000   
50%        3859.000000        1084.500000  128.000000        360.000000   
75%        5852.500000        2253.250000  170.000000        360.000000   
max       81000.000000       33837.000000  600.000000        480.000000   

       Credit_History  
count      480.000000  
mean         0.854167  
std          0.353307  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [56]:
Education_loan_data=data[['Education', 'Loan_Status']].groupby(['Education', 'Loan_Status']).agg(Count = ('Loan_Status','count')).reset_index()
Gender_loan_data=data[['Gender', 'Loan_Status']].groupby(['Gender', 'Loan_Status']).agg(Count = ('Loan_Status','count')).reset_index()
Self_Employed_loan_data=data[['Self_Employed', 'Loan_Status']].groupby(['Self_Employed', 'Loan_Status']).agg(Count = ('Loan_Status','count')).reset_index()
Credit_History_loan_data=data[['Credit_History', 'Loan_Status']].groupby(['Credit_History', 'Loan_Status']).agg(Count = ('Loan_Status','count')).reset_index()


for col in ['Education','Gender','Self_Employed','Credit_History']:
  print('\n\n')
  df=col+'_loan_data'
  df = pd.DataFrame(eval(df))
  tricolor = ['#1f77b4', '#ff7f0e', '#f37736']
  col_label = col.replace("_", " " )
  fig1=px.bar(data_frame=df, x=col, y='Count', color='Loan_Status', barmode='group', color_discrete_sequence=tricolor, text_auto=True, template='simple_white')

  fig1.update_layout(yaxis_range=[0,350])  
  fig1.update_yaxes(showgrid=True)
  fig1.update_xaxes(showgrid=True)   
    
  fig1.update_layout(height=400,width=800,
      title=f'<b>{col_label} Vs Loan Status</b>',
      font_family="Times New Roman",
      title_font_family="Times New Roman",
      title_font_color="#000000",
      title_font_size=20,
      title_xanchor="center",
      title_x=0.5,
      xaxis_title=f"<b>{col_label}</b>",
      yaxis_title="<b>Count</b>",
      legend_title='<b>Loan Status</b>',
      legend_title_font_color="#000000",
      plot_bgcolor ='#ffffff'
  )


  fig1.show()
    
# We see that features Gender, Education and Self_Employed do not have significant impact on the Loan Status. 
# However, Credit history has a significant impact on the approval status of a loan.

In [58]:
gender_loan_data = data[['Gender', 'Loan_Status']].groupby(['Gender', 'Loan_Status']).size().reset_index(name='Count')
fig1 = px.bar(gender_loan_data, x='Gender', y='Count', color='Loan_Status', barmode='group')
fig1.show()

#We group the data by Gender and Loan_Status, and count the number of observations in each group.
#From the bar graph, we can interpret the following:
#The number of males who got the loans approved are 278 whereas 116 were rejected.
#The number of females who got the loans approved are 54 whereas 32 were rejected.
#Males have a higher loan approval rate as compared to females.

In [59]:
fig2 = px.scatter(data, x='ApplicantIncome', y='LoanAmount', color='Loan_Status')
fig2.show()


# Relationship between Applicant Income and Loan Amount
# The scatter plot shows whether there is any correlation between Applicant Income and Loan Amount.
# It also shows how the loan status varies across different values of the variables.

In [60]:
# For each property area, what is the average loan amount of the people of that type?
average_loan = data.groupby('Property_Area')['LoanAmount'].mean()
print(average_loan)
# We see that people from Rural have the highest average loan amount.

Property_Area
Rural        155.640288
Semiurban    146.057592
Urban        132.946667
Name: LoanAmount, dtype: float64


In [61]:
# How many applicants based on gender are self employed or not? 
count = data.groupby(['Gender', 'Self_Employed'])['Loan_ID'].count()
print(count)

Gender  Self_Employed
Female  No                74
        Yes               12
Male    No               340
        Yes               54
Name: Loan_ID, dtype: int64


We can compare the count of loan applicants for each combination of gender and self-employment status.
The output shows that there are 340 male applicants who are not self-employed, 12 female applicants who are self-employed, and 15 female applicants who are not self-employed. 

In [62]:
# How income differs based on sex?
income_gender = data.groupby('Gender')['ApplicantIncome'].mean().reset_index()
print(income_gender)

   Gender  ApplicantIncome
0  Female      4968.593023
1    Male      5450.588832


In [68]:
# How does the credit history of loan applicants affect the loan approval rate? 
credit_loan_status = pd.crosstab(data['Credit_History'], data['Loan_Status'], margins=True, margins_name='Total')
print(credit_loan_status)

Loan_Status       N    Y  Total
Credit_History                 
0.0              63    7     70
1.0              85  325    410
Total           148  332    480


In [76]:
# How does marriage and dependents affect the property?
dependent_marriage = data.groupby(['Married', 'Dependents', 'Property_Area'])['Loan_ID'].count().reset_index()
print(dependent_marriage)

   Married Dependents Property_Area  Loan_ID
0       No          0         Rural       43
1       No          0     Semiurban       57
2       No          0         Urban       43
3       No          1         Rural        3
4       No          1     Semiurban        6
5       No          1         Urban        6
6       No          2         Rural        4
7       No          2     Semiurban        2
8       No          2         Urban        1
9       No         3+         Rural        3
10      No         3+     Semiurban        1
11     Yes          0         Rural       41
12     Yes          0     Semiurban       56
13     Yes          0         Urban       34
14     Yes          1         Rural       12
15     Yes          1     Semiurban       26
16     Yes          1         Urban       27
17     Yes          2         Rural       19
18     Yes          2     Semiurban       28
19     Yes          2         Urban       31
20     Yes         3+         Rural       14
21     Yes